sw/device/lib/dif/README.md

```diff
@@ -12,3 +12,54 @@ libraries.
 There is one DIF library per hardware IP, and each one contains the DIFs
 required to actuate all of the specification-required functionality of the
 hardware they are written for.
+
+## DIF Style Guide
+
+DIFs are very low-level software, so they have a more rigorous coding style than
+other parts of the codebase.
+
+DIFs should follow the [OpenTitan C/C++ style
+guide](https://docs.opentitan.org/doc/rm/c_cpp_coding_style/) where it does not
+contradict with the guidelines below.
+
+The guidelines below apply to writing DIFs, and code should be written in a
+similar style to the existing DIF libraries in this directory.
+
+### DIF Library Guidance
+
+* DIF libraries must be written in C.
+* DIF libraries can only depend on the headers and libraries in the
+  `sw/device/lib/base` directory.
+* DIF libraries must not depend on other DIF libraries. Exercising DIF
+  functionality may require an environment set up using another DIF library, but
+  DIFs must not call DIFs in other DIF libraries.
+
+### DIF Guidance
+
+* DIF declarations must match their definitions exactly.
+  * Scalar arguments should not be marked `const` in DIF signatures.
+
+* DIFs must use enum return codes rather than booleans for reporting errors. If
+  a DIF can error or produce a value, it should return an enum value, and use an
+  out-parameter for returning the value.
+
+* DIFs must check their arguments against preconditions. DIFs must "early
+  return" with an error code if their arguments fail the following checks:
+  * DIFs must ensure their pointer arguments are non-null, unless that pointer
+    is for an optional out-parameter.
+  * DIFs must ensure, if they only accept a subset of an enum, that the argument
+    is within that subset. However, DIFs may assume that the argument is one of
+    the enum variants.
+  * DIFs must not have side-effects before an "early return". Side-effects
+    include (but are not limited to) writing to memory, including memory-mapped
+    hardware, and modifying processor CSRs.
+
+* Switch statements in DIFs must always have a default case, including when
```

Comment:
Unless you also enable `-Wswitch-enum` I think you're slightly weakening the compilers ability to warn for coding errors, in cases where you intend but fail to cover the whole enum.

I suspect enabling `-Wswitch-enum` would be too noisy, and so I'm not convinced this guidance is the right choice.

Reply:
The intention of `default` is to catch two cases:
- when a programmer has passed a representable value that does not correspond to an enum variant value.
- when someone has glitched the core, and a valid value was passed, but the hardware memory now contains an invalid but representable value.

`-Wswitch-enum` will give errors even if there is a `default`, but does not check for reachability (so if one variant is unreachable due to a prior guard, it will still error if there is no explicit `case` to the variant). I think there could be a case for enabling `-Wswitch-enum` for the DIFs, but I would like to see more DIF code before we do so. 

Personally, I would like to see as much help as possible for programmers who intended, but have not managed, to cover the whole enum. This I would like to see balanced with being prudent about glitching and invalid variants.

Reply:
I believe in the defensive benefit of default here.